source/diffHandler.py

```diff
@@ -0,0 +1,192 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2020 Bill Dengler
+
+import config
+import globalVars
+import os
+import struct
+import subprocess
+import sys
+from abc import abstractmethod
+from baseObject import AutoPropertyObject
+from difflib import ndiff
+from logHandler import log
+from textInfos import TextInfo, UNIT_LINE
+from threading import Lock
+from typing import List
+
+
+class DiffAlgo(AutoPropertyObject):
+	@abstractmethod
+	def diff(self, newText: str, oldText: str) -> List[str]:
+		raise NotImplementedError
+
+	@abstractmethod
+	def _getText(self, ti: TextInfo) -> str:
+		raise NotImplementedError
+
+
+class DiffMatchPatch(DiffAlgo):
+	"""A character-based diffing approach, using the Google Diff Match Patch
+	library in a proxy process (to work around a licence conflict).
+	"""
+	#: A subprocess.Popen object for the nvda_dmp process.
+	_proc = None
+	#: A lock to control access to the nvda_dmp process.
+	#: Control access to avoid synchronization problems if multiple threads
+	#: attempt to use nvda_dmp at the same time.
+	_lock = Lock()
+
+	def _initialize(self):
+		"Start the nvda_dmp process if it is not already running."
+		with DiffMatchPatch._lock:
+			if not DiffMatchPatch._proc:
+				log.debug("Starting diff-match-patch proxy")
+				if hasattr(sys, "frozen"):
+					dmp_path = (os.path.join(globalVars.appDir, "nvda_dmp.exe"),)
+				else:
+					dmp_path = (sys.executable, os.path.join(
+						globalVars.appDir, "..", "include", "nvda_dmp", "nvda_dmp.py"
+					))
+				DiffMatchPatch._proc = subprocess.Popen(
+					dmp_path,
+					creationflags=subprocess.CREATE_NO_WINDOW,
+					bufsize=0,
+					stdin=subprocess.PIPE,
+					stdout=subprocess.PIPE
+				)
+
+	def _getText(self, ti: TextInfo) -> str:
+		return ti.text
+
+	def diff(self, newText: str, oldText: str) -> List[str]:
+		try:
+			if not newText and not oldText:
+				# Return an empty list here to avoid exiting
+				# nvda_dmp uses two zero-length texts as a sentinal value
+				return []
+			self._initialize()
+			with DiffMatchPatch._lock:
+				old = oldText.encode("utf-8")
+				new = newText.encode("utf-8")
+				# Sizes are packed as 32-bit ints in native byte order.
+				# Since nvda and nvda_dmp are running on the same Python
+				# platform/version, this is okay.
+				tl = struct.pack("=II", len(old), len(new))
+				DiffMatchPatch._proc.stdin.write(tl)

```

Comment:
- `_terminate` could occur between initialize and this lock acquisition
- If `_terminate` has been called (and thus `_proc` is None) it is probably an error for `_initialize` to be called again. This may happen if the users has a terminal focused while exiting NVDA for instance.

Reply:
This still needs to be addressed.

Reply:
How do you suggest addressing this beyond what's already there? As of the latest commit:

* If `_terminate` is called, `_proc` is no longer set to `None`.
* If `_initialize` is called after `_terminate`, `_proc` is not `None` and so nvda_dmp is not initialized.

Reply:
Imagine the following "NVDA is shutting down, there are two threads, one (T1) is trying to call diff, the other (T2) is trying to call terminate".

There are two ways I this will currently cause problems, here is the first.
- T1 `diff` is called first
- T1 the lock in `_initialize` is acquired, the T2 is blocked in `_terminate` trying to acquire the lock.
- T1 ` _initialize` finishes, the lock is released.
- Before the lock is reacquired from T1 to perform the diff, T2 acquires the lock from the `_terminate` method.
- Now T1 is blocked in the body of the `diff` method.
- In T2, `_proc` is set to `None`, `_terminate` finishes
- T1 can continue through the `diff` method, however it tries to use `_proc` which is now set to `None`.

Here is the second:
- `_terminate` is called first, acquires the lock,
- `diff` goes into `_initialize` and is blocked trying to acquire the lock
- `_terminate` finishes, setting `_proc` to `None`, releases the lock
- the `diff` thread then re-initializes `_proc` (restarting the process)

I'd suggest either:
- wrapping the call to initialize in the context of an acquired lock (from the `diff` method)
- OR rechecking that `_proc` is still set after the lock is acquired in the `diff` method